### PR TITLE
Guided tour for Editor

### PR DESCRIPTION
**Pull request by janithRS**
## Purpose
Give a new user a guide through the process of implementing a simple siddhi app and simulating it from the event simulator using both the source view and the design view.

## Goals
Display step by step hints using the Enjoyhint library to the user what needs to be done.

## Preview
peek 2019-03-06 12-15

### CODE REVIEW

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 1]
```diff
@@ -0,0 +1,888 @@
...
+                        var fileToBeChecked = "configName=" + btoa(Constants.TEMP_FILE + '.siddhi');
+
+                        $.ajax({
+                            url: checkFileURL,
+                            type: "POST",
+                            contentType: "text/plain; charset=utf-8",
+                            data: fileToBeChecked,
+                            async: false,
+                            success: function (response) {
+                                if(response.exists === true){
```
Comment (lasanthaS) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Instead use `if (response.exists)`

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+                                    Constants.TEMP_FILE = Constants.TEMP_FILE.slice(0, 14);
```
Comment (lasanthaS) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Assign to local variables

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 3]
```diff
@@ -0,0 +1,888 @@
...
+                },
+                {
+                    'next #stream': 'This is a <b class="lime-text">Stream </b>component. you need to use this to ' +
+                        'generate the output. Click <b class="lime-text">Next</b>.',
+                    'showSkip': false,
+                    'showNext': true,
+                    onBeforeStart: function () {
+                        $('#stream').removeClass('stream-drag');
+                        Constants.CURRENT_STEP = instance.getCurrentStep();
+
```
Comment (lasanthaS) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Remove unnecessary new lines.

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 4]
```diff
@@ -0,0 +1,888 @@
...
+                        $('#tool-group-Queries').find('.tool-group-body').css('display', 'none');
+                        $('#tool-group-Functions').find('.tool-group-body').css('display', 'none');
+                        $("#tool-group-I\\/O").find('.tool-group-body').css('display', 'none');
+                        $('#trigger').removeClass('trigger-drag');
+                        $('#partition').removeClass('partition-drag');
+                        $('#stream').addClass('stream-drag');
+                        Constants.CURRENT_STEP = instance.getCurrentStep();
+
+                        setTimeout(function () {
+                            var Interval = null;
```
Comment (lasanthaS) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Use camel-case for variable names.

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 5]
```diff
@@ -0,0 +1,888 @@
...
+                if (self.tabList.length === 1 && tab._title === "welcome-page") {
+                    self.showModel();
+                } else {
+                    DesignViewUtils.prototype.errorAlert(Constants.ERROR_TEXT);
+                }
+            });
+
+            //function for the complete tour
+            guideModal.find("button").filter("#fullGuide").click(function() {
+
```
Comment (lasanthaS) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Remove blank line

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 6]
```diff
@@ -0,0 +1,888 @@
...
+                //unbinding the click events from the previous click
+                $('#fullGuide').off('click');
+                $('#simulationGuide').off('click');
+                $('#sampleGuide').off('click');
+
+                hintInstance.setScript(self.completeGuide);
+                hintInstance.runScript();
+
+                instance = hintInstance;
```
Comment (lasanthaS) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Why assigning `hintInstance` to `instance`? Can't we use single variable for this?

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 7]
```diff
@@ -0,0 +1,888 @@
...
+                hintInstance.setScript(self.sampleGuide);
+                hintInstance.runScript();
+
+                instance = hintInstance;
+                callback();
+            });
+
+            $('.enjoyhint_close_btn').click(function () {
+                switch (Constants.CURRENT_STEP) {
+                    case 7  :   $('#stream').addClass('stream-drag');
```
Comment (lasanthaS) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Format code like,

```js
switch(...) {
    case 7:
        // code goes here
    // ...
}
```

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/css/customEnjoyhint.css [hunk 8]
```diff
@@ -0,0 +1,27 @@
+.enjoyhint {
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/css/customEnjoyhint.css]: Added

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/css/customEnjoyhint.css [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+.tour-guide-button {
+    color: #f1f6fa;
+    font-size: 1.286em;
+    background-color: #f17b31;
+    margin-bottom: 25px;
+    z-index: 1;
+    position: relative;
+}
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/css/customEnjoyhint.css]: Added

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/js/menu-bar/tools-menu.js [hunk 10]
```diff
@@ -110,6 +110,24 @@ define(([],function (){
                     }
                 },
                 disabled: false
+            },
+            {
+                id: "hintTour",
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/js/menu-bar/tools-menu.js]: Changed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/js/workspace/workspace.js [hunk 11]
```diff
@@ -752,6 +757,11 @@ define(['ace/ace', 'jquery', 'lodash', 'log','dialogs','./service-client','welco
             // Open Sample file open dialog
             app.commandManager.registerHandler('open-sample-file-open-dialog', this.openSampleFileOpenDialog, this);
             app.commandManager.registerHandler('query-store', this.openQueryStore, this);
+
+            // //Run hint tour
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/js/workspace/workspace.js]: Removed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/index.html [hunk 12]
```diff
@@ -28,6 +30,11 @@
     <link href="/editor/commons/css/simulator.css" rel="stylesheet" type="text/css"/>
     <link rel="stylesheet" type="text/css" href="/editor/commons/css/notification.css"/>
 
+    <!--Enjoy Hint CSS-->
+
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/index.html]: Removed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/index.html [hunk 13]
```diff
@@ -1525,6 +1532,34 @@ <h4 class="modal-title file-dialog-title" id="aboutDockerExportLabel">Export as
 
 <iframe id="frm-download" style="display: none;"></iframe>
 
+<div class='modal fade' id='guideDialog' tabindex='-1' role='dialog' aria-tydden='true'>
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/index.html]: Changed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/design-view/form-builder/forms/sink-form.js [hunk 14]
```diff
@@ -310,7 +310,8 @@ define(['log', 'jquery', 'lodash', 'mapAnnotation', 'payloadOrAttribute', 'jsonV
 
                 //onclick of submit
                 var submitButtonElement = $(formContainer).find('#btn-submit')[0];
-                submitButtonElement.addEventListener('click', function () {
+                $(formContainer).on('click', '#btn-submit', function () {
+                // submitButtonElement.addEventListener('click', function () {
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/design-view/form-builder/forms/sink-form.js]: Removed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/design-view/form-builder/forms/stream-form.js [hunk 15]
```diff
@@ -118,8 +118,8 @@ define(['require', 'log', 'jquery', 'lodash', 'attribute', 'jsonValidator', 'con
             self.formUtils.checkPredefinedAnnotations(checkedAnnotations);
 
             //submit button action
-            var submitButtonElement = $(formContainer).find('#btn-submit')[0];
-            submitButtonElement.addEventListener('click', function () {
+            $(formContainer).on('click', '#btn-submit', function () {
+                //.submitButtonElement.addEventListener('click', function () {
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/design-view/form-builder/forms/stream-form.js]: Removed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide-constants.js [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+            'insert into TotalProductionStream;\n',
+        INSERT_STRING: 'define stream SweetProductionStream (name string, amount long);',
+        TAB_KEYCODE : 9,
+        ERROR_TEXT: "Please close all tabs except welcome-page to start the guide. You can use close all button on the" +
+            " right corner of the screen",
+        CURRENT_STEP: null
+    };
+
+    return constants;
+});
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide-constants.js]: Added

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/main.js [hunk 17]
```diff
@@ -141,6 +144,14 @@ define(['require', 'log', 'jquery', 'lodash', 'backbone', 'menu_bar', 'tool_bar'
                     this.workspaceManager.displayInitialTab();
                 },
 
+                runInitialGuide: function (){
+                    var isFreshUser = this.browserStorage.get('guideFileNameIncrement');
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/main.js]: Changed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 18]
```diff
@@ -0,0 +1,888 @@
...
+
+        //enjoyhint instance to be used in the methods of this class
+        var instance = null;
+
+        /**
+         * Arg: application instance
+         */
+
+        var Guide = function (application) {
+
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Removed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 19]
```diff
@@ -0,0 +1,888 @@
...
+                        'name, and set of defined attributes with specific data types defining its schema. ' +
+                        '<br>Click <b class="lime-text">Next</b> to continue.',
+                    'margin': 200,
+                    'showSkip': false,
+                    onBeforeStart: function () {
+                        var activeTab = self.app.tabController.getActiveTab();
+                        var editor = activeTab.getSiddhiFileEditor().getSourceView().getEditor();
+                        var aceEditor = self.app.tabController.getActiveTab().getSiddhiFileEditor().getSourceView().getEditor();
+                        editor.session.insert(aceEditor.getCursorPosition(), Constants.INSERT_STRING);
+                        Constants.CURRENT_STEP = null;
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Changed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 20]
```diff
@@ -0,0 +1,888 @@
...
+                {
+                    'keyCode': Constants.TAB_KEYCODE,
+                    'selector': '#saveName',
+                    'event': 'key',
+                    'description': 'Your file name is this. Press <b class="lime-text">Tab </b>key to continue.',
+                    'showNext': false,
+                    'showSkip': false,
+                    'timeout': 200,
+                    onBeforeStart: function () {
+                        Constants.FILE_INCREMENT = browserStorage.get('guideFileNameIncrement');
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Changed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js (hunk 1 above)
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Changed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js (hunk 2 above)
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: changed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js (hunk 3 above)
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Removed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js (hunk 4 above)
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Changed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js (hunk 5 above)
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Removed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js (hunk 6 above)
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Changed to use a single variable

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js (hunk 7 above)
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Formatted

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/js/menu-bar/tools-menu.js [hunk 21]
```diff
@@ -110,6 +110,24 @@ define(([],function (){
...
+                label: "Tour guide",
+                command: {
+                    id: "hint-tour",
+                    labels:{
+                        mac: {
+                            key: "",
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/js/menu-bar/tools-menu.js]: Added Ctrl+Shift+H

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/main.js [hunk 22]
```diff
@@ -141,6 +144,15 @@ define(['require', 'log', 'jquery', 'lodash', 'backbone', 'menu_bar', 'tool_bar'
                     this.workspaceManager.displayInitialTab();
                 },
 
+                runInitialGuide: function (){
+
+                    var isFreshUser = (this.browserStorage.get('guideFileNameIncrement') === null);
+                    if(isFreshUser) {
+                        this.workspaceManager.runGuide();
```
Comment (dnwick) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/main.js]: Use the already created object to run the guide without using workspace

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/css/customEnjoyhint.css [hunk 23]
```diff
@@ -0,0 +1,45 @@
+/**
```
Comment (dnwick) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/css/customEnjoyhint.css]: change the css name to guide.css

Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/commons/css/customEnjoyhint.css]: Changed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/design-view/form-builder/forms/window-filter-projection-query-form.js [hunk 24]
```diff
@@ -1124,7 +1124,8 @@ define(['require', 'log', 'jquery', 'lodash', 'querySelect', 'queryOutputInsert'
 
                 // 'Submit' button action
                 var submitButtonElement = $(formContainer).find('#btn-submit')[0];
-                submitButtonElement.addEventListener('click', function () {
+                $(formContainer).on('click', '#btn-submit', function () {
+                // submitButtonElement.addEventListener('click', function () {
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/design-view/form-builder/forms/window-filter-projection-query-form.js]: Removed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide-constants.js [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * Constants to be used by the siddhi editor tour guide
+ */
+define(function () {
```
[the remark this answers is not in the document]
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide-constants.js]: Changed

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/main.js (hunk 22 above)
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/main.js]: Changed the object to the existing

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/index.html [hunk 26]
```diff
@@ -1759,6 +1793,8 @@ <h4>Return Attributes: </h4>
             datatables: "commons/lib/data-table/jquery.dataTables.min",
             datatables_bootstrap: "commons/lib/data-table/dataTables.bootstrap.min",
             datatables_wso2: "commons/lib/theme-wso2-2.0.0/extensions/datatables/js/dataTables.wso2",
+            enjoyhint: "commons/lib/enjoyhint/enjoyhint.min",
+            guide: "js/guide/guide",
```
Comment (dnwick) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/index.html]: move this under custom modules

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 27]
```diff
@@ -0,0 +1,873 @@
...
+            guideModal.find("button").filter("#sampleGuide").unbind().click(function() {
+                //unbinding the click events from the previous click
+                instance = new EnjoyHint({});
+                $('#fullGuide').off('click');
+                $('#simulationGuide').off('click');
+                $('#sampleGuide').off('click');
+                instance.setScript(self.sampleGuide);
+                instance.runScript();
+                callBack();
+                test();
```
Comment (dnwick) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: what do u mean by test ?

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js [hunk 28]
```diff
@@ -0,0 +1,873 @@
...
+                        case 21 :
+                            $('#sink').addClass('sink-drag');
+                            break;
+                        case 22 :
+                            $('#source').addClass('source-drag ');
+                            break;
+                        default :
+                            break;
+                    }
+
```
Comment (dnwick) [inline components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: remove unnecessary new lines

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/index.html (hunk 26 above)
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/index.html]: Moved

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js (hunk 27 above)
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Changed the name to guideCloseCallback

File: components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js (hunk 28 above)
Comment (janithRS) [reply components/org.wso2.carbon.siddhi.editor.core/src/main/resources/web/js/guide/guide.js]: Removed

**Pull request merged**